### PR TITLE
Use celery to schedule periodic tasks

### PR DESCRIPTION
**Pull request by Iain-S**
## Summary

As we add more periodic tasks (e.g. if we add usage forecasting in addition to the existing abolishments and daily emails), switching to a more widely used task management option makes sense.

Celery is a task queue, which requires a broker to mediate between clients and workers and, optionally, a backend for result storage. I have used Redis as the broker as it was easier to install than Rabbit MQ. Since we don't do anything with the results of completed tasks, I have omitted a backend but we could use Redis or Postgres if we need one in the future.

## ToDo

- [x] Fix the error that can occur if we try to insert a daily summary email for subscription `UUID(int=0)`.
- [x] Add `Europe/London` timezone?
- [ ] ~~Shut down redis and celery on exit (particularly if using the runtests script.~~
- [x] Set up Celery logging (currently logging to file)

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -26,6 +26,11 @@ jobs:
           POSTGRES_PASSWORD: password
         ports:
           - 5432:5432
+      redis:
+        image: redis
+        ports:
+          # Opens tcp port 6379 on the host and service container
+          - 6379:6379
```
Comment (Iain-S) [inline .github/workflows/test.yml:33]: Need Redis service to run celery tests.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -44,16 +49,7 @@ jobs:
...
-
-      # - run: poetry run coverage xml
-      # - run: poetry run genbadge coverage --input-file coverage.xml --local
-      # - uses: LanceMcCarthy/Action-AzureBlobUpload@v1.11.0
-      #   name: Coverage Azure Blob Upload
-      #   with:
-      #     connection_string: ${{ secrets.AZURE_BLOB_CONNECTION_STR }}
-      #     container_name: rctabcoveragecontainer
-      #     source_folder: coverage-badge.svg
+        run: CELERY_RESULT_BACKEND="redis://localhost:6379/0" scripts/runtests.sh -d -c main
```
Comment (Iain-S) [inline .github/workflows/test.yml:52]: The Celery tests only run when this env var is set.

File: Dockerfile [hunk 3]
```diff
@@ -1,18 +1,27 @@
-FROM docker.io/tiangolo/uvicorn-gunicorn-fastapi:python3.10-slim
+FROM docker.io/tiangolo/uvicorn-gunicorn-fastapi:python3.11-slim
```
Comment (Iain-S) [inline Dockerfile:1]: Note, update to 3.11

File: docs/content/logging.md [hunk 4]
```diff
@@ -103,12 +103,12 @@ You can further filter the log messages on the Azure portal.
...
-| extend logger_name = tostring(customDimensions.logger_name)
-| extend module = tostring(customDimensions.module)
-| extend line_number = tostring(customDimensions.lineNumber)
+AppTraces
+| extend logger_name = tostring(Properties.logger_name)
+| extend module = tostring(Properties.module)
+| extend line_number = tostring(Properties.lineNumber)
 | where module == "main"
-| where message contains "Starting server"
+| where Message has "Starting server"
```
Comment (Iain-S) [inline docs/content/logging.md:111]: See <https://learn.microsoft.com/en-us/azure/azure-monitor/app/convert-classic-resource#table-structure>

File: rctab/tasks.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+my_logger = logging.getLogger(__name__)
+
+CELERY_BROKER_URL: Final = "redis://localhost:6379/0"
+
+celery_app = Celery(
+    "rctab.tasks",
+    broker=CELERY_BROKER_URL,
+)
+
+celery_app.conf.timezone = "Europe/London"
```
Comment (Iain-S) [inline rctab/tasks.py:40]: To account for British Summer Time.

File: rctab/tasks.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    sender.add_periodic_task(
+        crontab(hour="16", minute="00"), run_send_summary_email.s()
+    )
+    sender.add_periodic_task(
+        crontab(hour="01", minute="00"), run_abolish_subscriptions.s()
+    )
+
+
+@after_setup_logger.connect
+def setup_logger(logger: Any, *_: Any, **__: Any) -> None:  # type: ignore[no-untyped-def]
```
Comment (Iain-S) [inline rctab/tasks.py:55]: This setup function and the one below allow logging to our centralised workspace.

**Review by joseph-palmer — APPROVED** [review]
🚀

**Pull request merged**